### ISSUE TITLE
JSDoc support for String array to Object key

### ISSUE THREAD

**Issue by imba97** [microsoft/TypeScript#45902]
# Suggestion

## 🔍 Search Terms

List of keywords you searched for before creating this issue. Write them down here so that others can find this suggestion more easily and help provide feedback.

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

JSDoc support for String array to Object key

## 📃 Motivating Example

```javascript
/**
 * @template T
 * @param {T[]} obj
 * @returns {Readonly<{[T]: number}>}
 */
function Flags(obj) {
  const _obj = {}
  _.forEach(obj, (key, index) => {
    _obj[key] = 1 << index
  })
  return Object.freeze(_obj)
}

const TEST = Flags([
  'A',
  'B',
  'C',
  'D'
])

// We want to

// TEST.
//     -----------
//     | A       |
//     | B       |
//     | C       |
//     | D       |
//     -----------
```

## 💻 Use Cases

In the above example, I can use bit enumeration

```javascript
/**
 * @param {number} enums
 * @param {number} flags
 */
function hasFlag(enums, flags) {
  return flags === (enums & flags)
}

cosnt test = TEST.A | TEST.C | TEST.D

console.log(hasFlag(test, TEST.A)) // true
console.log(hasFlag(test, TEST.B)) // false
console.log(hasFlag(test, TEST.C)) // true

```

**Comment by andrewbranch** [comment]
https://www.typescriptlang.org/play?ts=4.5.0-dev.20210910&filetype=js#code/PQKhCgAIUgBAXApgWwA4BsCGTIG8BOimAJgPYB26AnpAM7z4CW5A5gNoC6AvpACpQxYqTPkzI8vHqQBGAKwFxC8AK75ytPACUiZSlQA8uNgGlIzPm3LLk0xPg4cAXJCs27XAHxcBwcADNlcgBjeEYKSAAxLBZaAAoZWQBKPChIIIp6SAB9BMgAXjxvSGyAOj9SfABRTCCAC3i5ABpIWIBrRCpm5mJEAA9kvI8U4uKcuTZ2qg58yABGSH19M3Ie3tSuRNSlVXJIAHk5RBCywkQAL0RYsaTwb3B09Xg+SoBlXhmozBjY0EF4KlQiDwD3oPBAwBabFSAHIAILQxowgBCCJhAGFUcVoQARaHgDiJTbgXivXglcBAA

**Comment by imba97** [comment]
> https://www.typescriptlang.org/play?ts=4.5.0-dev.20210910&filetype=js#code/PQKhCgAIUgBAXApgWwA4BsCGTIG8BOimAJgPYB26AnpAM7z4CW5A5gNoC6AvpACpQxYqTPkzI8vHqQBGAKwFxC8AK75ytPACUiZSlQA8uNgGlIzPm3LLk0xPg4cAXJCs27XAHxcBwcADNlcgBjeEYKSAAxLBZaAAoZWQBKPChIIIp6SAB9BMgAXjxvSGyAOj9SfABRTCCAC3i5ABpIWIBrRCpm5mJEAA9kvI8U4uKcuTZ2qg58yABGSH19M3Ie3tSuRNSlVXJIAHk5RBCywkQAL0RYsaTwb3B09Xg+SoBlXhmozBjY0EF4KlQiDwD3oPBAwBabFSAHIAILQxowgBCCJhAGFUcVoQARaHgDiJTbgXivXglcBAA

Thank you sir
Is it available only in v4.5.0?

**Comment by imba97** [comment]
Thank you very much
